### PR TITLE
Ensure inverse payout curve does not go over oracle's maximum BTC price

### PR DESCRIPTION
**Pull request by luckysori**
Fixes to #1566.

@bonomat: Maybe you want to take this over, I just managed to reproduce the cause for the error seen in #1566 using the data from the logs. My understanding is that we just need to ensure that the price ranges can never surpass `2^20 - 1`. And that we should write a proptest to verify this too.

### CODE REVIEW

**Comment by bonomat** [comment]
Happy to take this over.

**Comment by bonomat** [comment]
Are you sure that it should be `<2^20 - 1` and not `< 2^20` ? Because if so, then all we need to do is set 
`BTCUSD_MAX_PRICE` to `1_048_574;` and not `1_048_575` which it is at the moment;

Do you have the test to replicate the error in https://github.com/get10101/10101/issues/1566 to verify this?

**Comment by luckysori** [comment]
> Are you sure that it should be `<2^20 - 1` and not `< 2^20` ? 

I think `2^20 - 1` is the _inclusive_ upper bound.

> Because if so, then all we need to do is set `BTCUSD_MAX_PRICE` to `1_048_574;` and not `1_048_575` which it is at the moment;

I think the test that I included in this PR generates an price range upper bound of `1_048_58X`. Does changing the constant also fix that?

> Do you have the test to replicate the error in #1566 to verify this?

It's in the first patch I pushed here.

File: crates/trade/src/cfd.rs [hunk 1]
```diff
@@ -8,7 +8,9 @@ use rust_decimal::Decimal;
 use rust_decimal_macros::dec;
 use std::ops::Neg;
 
-pub const BTCUSD_MAX_PRICE: u64 = 1_048_575;
+// TODO: is this correct? according to luckysori this should be 2^20 - 1 (i.e. 1_048_574) and not
+// 2^20 1_048_575) as it used to be
+pub const BTCUSD_MAX_PRICE: u64 = 1_048_574;
```
Comment (bonomat) [inline crates/trade/src/cfd.rs:13]: Don't mind this change :/

**Comment by bonomat** [comment]
> > Are you sure that it should be `<2^20 - 1` and not `< 2^20` ?
> 
> I think `2^20 - 1` is the _inclusive_ upper bound.

Right, not sure what I thought... `2^20 - 1=1048575` which is what we have already. 

Your test check though if all values are `<` and not `<=`. I think it should be `<=`. 

> 
> > Because if so, then all we need to do is set `BTCUSD_MAX_PRICE` to `1_048_574;` and not `1_048_575` which it is at the moment;
> 
> I think the test that I included in this PR generates an price range upper bound of `1_048_58X`. Does changing the constant also fix that?

This is what I believe I've fixed in https://github.com/get10101/10101/pull/1568/commits/3547e04e24ec7929cd42dfac4975c50e051d4fb0

> 
> > Do you have the test to replicate the error in #1566 to verify this?
> 
> It's in the [first patch](https://github.com/get10101/10101/pull/1568/files#diff-cc998bd0f27ef868e570ee1003700c0654e29674b57337885276ae3f621346daR121) I pushed here.

See comment above, I think it should be `<=` then it works

**Review by holzeis — APPROVED** [review]
LGTM 👍

**Pull request merged**